pkg/attribute/protoBag.go

```diff
@@ -0,0 +1,159 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attribute
+
+import (
+	mixerpb "istio.io/api/mixer/v1"
+)
+
+// ProtoBag implements the Bag interface on top of an Attributes proto.
+type ProtoBag struct {
+	proto          *mixerpb.Attributes
+	globalDict     map[string]int32
+	globalWordList []string
+	messageDict    map[string]int32
+}
+
+func NewProtoBag(proto *mixerpb.Attributes, globalDict map[string]int32, globalWordList []string) *ProtoBag {
+	return &ProtoBag{
+		proto:          proto,
+		globalDict:     globalDict,
+		globalWordList: globalWordList,
+	}
+}
+
+// Get returns an attribute value.
+func (pb *ProtoBag) Get(name string) (interface{}, bool) {
+	index, ok := pb.globalDict[name]
+	if !ok {
+		md := pb.getMessageDict()
+
+		index, ok = md[name]
+		if !ok {
+			return nil, false
+		}
+	}
+
+	var value interface{}
+
+	value, ok = pb.proto.Strings[index]
+	if ok {
+		return value, true
+	}
+
+	value, ok = pb.proto.Int64S[index]
+	if ok {
+		return value, true
+	}
+
+	value, ok = pb.proto.Doubles[index]
+	if ok {
+		return value, true
+	}
+
+	value, ok = pb.proto.Bools[index]
+	if ok {
+		return value, true
+	}
+
+	value, ok = pb.proto.Timestamps[index]
+	if ok {
+		return value, true
+	}
+
+	value, ok = pb.proto.Durations[index]
+	if ok {
+		return value, true
+	}
+
+	value, ok = pb.proto.Bytes[index]
+	if ok {
+		return value, true
+	}
+
+	value, ok = pb.proto.StringMaps[index]
+	if ok {
+		return value, true
+	}
+
+	return nil, false
+}
+
+// Names return the names of all the attributes known to this bag.
+func (pb *ProtoBag) Names() []string {
+	var names []string
+	for k := range pb.proto.Strings {
+		name, _ := lookup(k, nil, pb.globalWordList, pb.proto.Words)
+		names = append(names, name)
+	}
+
+	for k := range pb.proto.Int64S {
+		name, _ := lookup(k, nil, pb.globalWordList, pb.proto.Words)
+		names = append(names, name)
+	}
+
+	for k := range pb.proto.Doubles {
+		name, _ := lookup(k, nil, pb.globalWordList, pb.proto.Words)
+		names = append(names, name)
+	}
+
+	for k := range pb.proto.Bools {
+		name, _ := lookup(k, nil, pb.globalWordList, pb.proto.Words)
+		names = append(names, name)
+	}
+
+	for k := range pb.proto.Timestamps {
+		name, _ := lookup(k, nil, pb.globalWordList, pb.proto.Words)
+		names = append(names, name)
+	}
+
+	for k := range pb.proto.Durations {
+		name, _ := lookup(k, nil, pb.globalWordList, pb.proto.Words)
+		names = append(names, name)
+	}
+
+	for k := range pb.proto.Bytes {
+		name, _ := lookup(k, nil, pb.globalWordList, pb.proto.Words)
+		names = append(names, name)
+	}
+
+	for k := range pb.proto.StringMaps {
+		name, _ := lookup(k, nil, pb.globalWordList, pb.proto.Words)
+		names = append(names, name)
+	}
+
+	return names
+}
+
+func (pb *ProtoBag) Done() {
+	// NOP
+}
+
+// Lazily produce the message-level dictionary
+func (pb *ProtoBag) getMessageDict() map[string]int32 {
```

Comment:
getMessageDict is called lazily. If no attributes are used that require the message-level dictionary, than getMessageDict never runs.